kafka-connect-aws-s3/src/main/scala/io/lenses/streamreactor/connect/aws/s3/sink/seek/IndexManager.scala

```diff
@@ -0,0 +1,184 @@
+package io.lenses.streamreactor.connect.aws.s3.sink.seek
+
+import cats.implicits._
+import com.typesafe.scalalogging.LazyLogging
+import io.lenses.streamreactor.connect.aws.s3.model.location.{RemoteS3PathLocation, RemoteS3RootLocation}
+import io.lenses.streamreactor.connect.aws.s3.model.{TopicPartition, TopicPartitionOffset}
+import io.lenses.streamreactor.connect.aws.s3.sink.{FatalS3SinkError, NonFatalS3SinkError, S3FileNamingStrategy, SinkError}
+import io.lenses.streamreactor.connect.aws.s3.storage.{FileDeleteError, FileLoadError, StorageInterface}
+
+class IndexManager(sinkName: String, maxIndexes: Int, fallbackSeeker: Option[LegacyOffsetSeeker])(implicit storageInterface: StorageInterface) extends OffsetSeeker with LazyLogging {
+
+  /**
+   * deletes all index files except for the one corresponding to topicPartitionOffset
+   *
+   * @param mostRecentIndexFile the latest offset successfully written
+   * @param topicPartition the topicPartition
+   */
+  def clean(mostRecentIndexFile: RemoteS3PathLocation, topicPartition: TopicPartition): Either[SinkError, Int] = {
+    val remoteS3PathLocation = mostRecentIndexFile.root().withPath(IndexFilenames.indexForTopicPartition(sinkName, topicPartition.topic.value, topicPartition.partition))
+    storageInterface.list(remoteS3PathLocation)
+      .leftMap(e => {
+        val logLine = s"Couldn't retrieve listing for (${mostRecentIndexFile.path}})"
+        logger.error("[{}] {}", sinkName, logLine, e.exception)
+        new NonFatalS3SinkError(logLine, e.exception)
+      })
+      .flatMap {
+        indexes: List[String] =>
+          val filtered = indexes.filterNot(_ == mostRecentIndexFile.path)
+          if (indexes.size > maxIndexes) {
+            logAndReturnMaxExceededError(topicPartition, indexes)
+          } else if (filtered.size == indexes.size) {
```

Comment:
will it ever get into this condition? if we have already a "index" file, will be deleted by the time this method executes?

Reply:
My hope is that it never gets into this condition, it is a precondition/safeguard only, this error message would be invaluable for troubleshooting.

Regarding the deleting, the current index will not be deleted until the next commit, so this will be safe, no?  Do you have concerns?